keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,402 @@
+
+- Everything which is not related to the container image pull procedure
+
+## Proposal / User Stories
+
+The overall implementation can be split up into three user stories.
+
+### CRI API enhancement
+
+The Kubernets runtime API `ImageService` has to be modified to provide an
+additional streaming RPC. This server-side stream provides continuously image
+pull related metadata during the whole image pull process. The new endpoint
+would be added in addition to the already available `PullImage` RPC and re-use
+the request data type:
+
+```protobuf
+service ImageService {
+    …
+
+    // PullImageStream pulls an image and provides continuous metadata during
+    // that time
+    rpc PullImageProgress(PullImageRequest) returns (stream PullImageProgressResponse) {}
+}
+
+enum PullImageState {
+    STARTED = 0;
+    PULLING = 1;
+    FAILED  = 2;
+    DONE    = 3;
+}
+
+message PullImageProgressResponse {
+    // The current state of the image pull proress
+    PullImageState state = 1;
+
+    // The overall size of the image in bytes
+    uint64 size = 2;
+
+    // The amount of data already retrieved in bytes
+    uint64 current_offset = 3;
+
+    // String indicating the reason for the current state, like a failure message
+    string reason = 4;
+}
+```
+
+The `PullImageState` provides richer information to the kubelet in which state
+the image pull currently resides as well as a fine granular error messages
+directly from the runtimes via the `reason` field.
+
+### Kubernetes API additions
+
+The already existing kubelet `imageManager` events `PullingImage`, `PulledImage`
+and `FailedToPullImage` will be re-used and extended with the additional
+metadata provided by the CRI. This means that the container will still reside in
+`ContainerStateWaiting` during the image pull, but the events regarding the
+image pull process will be thrown periodically. The amount of how often the
+kubelet throws this event should be configurable in a range between 1s and 1min.
+
+To reduce the impact of the added feature, a kubelet configuration option will
+be added which is (for now) disabled per default. This option completely avoids
+calling the new gRPC API and falls back to the current `PullImage` RPC.
+
+### CLI modifications
+
+The command line interface of Kubernetes (`kubectl`) does right now provide all
+necessary features to display the additional pull progress events, but the
+overall usability is not satisfying at all.
+
+To improve the amount of information when calling `kubectl get pods`, we will
+now take the thrown events into account as well. To achieve that and stay
+backwards compatible, a new output variant `events` will be added to the
+`kubectl get` flag `-o/--output`. If the flag is set, `kubectl` will map
+(manually implemented) relevant information into the default (table) output in a
+similar fashion to the `-o wide`. For example, the output for pods pulling a
+container image will now look like:
+
+```
+> kubectl get pods -o events
```

Comment:
we do have `kubectl get events --field-selector involvedObject.kind=Pod,involvedObject.name=pod1` for now.

Reply:
And `kubectl describe` could also show related events.

Reply:
Do you suggest in not modifying the command line at all? :)

Reply:
The problem could be solved via the CRD, too.

Reply:
Personally speaking, it's better to show events by specified command. `-o events` is fine to me.